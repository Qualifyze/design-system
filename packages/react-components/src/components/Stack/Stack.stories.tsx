import React from 'react';

import { StoryBackground } from '../private/StoryBackground/StoryBackground';
import { Box } from '../Box/Box';

import { Stack } from './Stack';

const boxStyles = { bc: 'white', br: '$8', height: '$space$32' };

export default {
  title: 'Components/Stack',
  component: Stack,
};

export const basic = () => (
  <StoryBackground>
    <Stack gap="24" align="end">
      <Box css={{ ...boxStyles, width: '$space$48' }} />
      <Box css={{ ...boxStyles, width: '$space$64' }} />
      <Box css={{ ...boxStyles, width: '$space$96' }} />
      <Box css={{ ...boxStyles, width: '$space$64' }} />
    </Stack>
  </StoryBackground>
);
