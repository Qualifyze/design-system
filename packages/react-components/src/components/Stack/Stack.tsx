import { styled } from '../../stitches.config';

export const Stack = styled('div', {
  boxSizing: 'border-box',
  position: 'relative',
  display: 'flex',
  flexDirection: 'column',

  variants: {
    align: {
      start: {
        alignItems: 'start',
      },
      center: {
        alignItems: 'center',
      },
      end: {
        alignItems: 'end',
      },
      stretch: {
        alignItems: 'stretch',
      },
    },
    gap: {
      0: {
        gap: 0,
      },
      4: {
        gap: '$4',
      },
      8: {
        gap: '$8',
      },
      12: {
        gap: '$12',
      },
      16: {
        gap: '$16',
      },
      24: {
        gap: '$24',
      },
      32: {
        gap: '$32',
      },
      48: {
        gap: '$48',
      },
      64: {
        gap: '$64',
      },
      96: {
        gap: '$96',
      },
    },
  },
  defaultVariants: {
    align: 'start',
    gap: '24',
  },
});
