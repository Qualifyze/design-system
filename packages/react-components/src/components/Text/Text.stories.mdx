import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import { Text } from './Text';

<Meta title="Components/Text" component={Text} />

# Text

The typographic heart of our design system.

<Canvas>
  <Story name="Text">
    <Text size="tiny" weight="regular">
      Text size tiny regular
    </Text>
    <Text size="tiny" weight="medium">
      Text size tiny medium
    </Text>
    <Text size="tiny" weight="strong">
      Text size tiny strong
    </Text>
    <Text size="small" weight="regular">
      Text size small regular
    </Text>
    <Text size="small" weight="medium">
      Text size small medium
    </Text>
    <Text size="small" weight="strong">
      Text size small strong
    </Text>
    <Text size="standard" weight="regular">
      Text size standard regular
    </Text>
    <Text size="standard" weight="medium">
      Text size standard medium
    </Text>
    <Text size="standard" weight="strong">
      Text size standard strong
    </Text>
    <Text size="large" weight="regular">
      Text size large regular
    </Text>
    <Text size="large" weight="medium">
      Text size large medium
    </Text>
    <Text size="large" weight="strong">
      Text size large strong
    </Text>
  </Story>
  <Story name="In paragraphs">
    <Text>
      MVP analytics iPad startup angel investor creative focus branding
      business-to-business holy grail crowdfunding marketing deployment beta.
      Branding client marketing funding sales disruptive direct mailing
      incubator. User experience product management venture beta scrum project
      niche market equity business-to-business. Crowdfunding buyer partner
      network.
    </Text>
    <Text>
      Business-to-business MVP metrics. Branding hackathon incubator social
      proof paradigm shift alpha. Incubator validation partner network burn rate
      mass market business-to-consumer research & development launch party value
      proposition monetization business-to-business advisor business model
      canvas partnership. Gamification scrum project branding ecosystem
      bootstrapping business plan first mover advantage customer iPad.
    </Text>
  </Story>
</Canvas>
