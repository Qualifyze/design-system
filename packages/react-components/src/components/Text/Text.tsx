import capsize from 'capsize';

import { styled } from '../../stitches.config';

// Font metrics for "Inter"
const fontMetrics = {
  capHeight: 2048,
  ascent: 2728,
  descent: -680,
  lineGap: 0,
  unitsPerEm: 2816,
};

const tinyStyles = capsize({
  capHeight: 8,
  lineGap: 10,
  fontMetrics,
});

const smallStyles = capsize({
  capHeight: 10,
  lineGap: 12,
  fontMetrics,
});

const standardStyles = capsize({
  capHeight: 12,
  lineGap: 18,
  fontMetrics,
});

const largeStyles = capsize({
  capHeight: 14,
  lineGap: 20,
  fontMetrics,
});

export const Text = styled('p', {
  // Reset
  lineHeight: '1',
  margin: '0',
  fontWeight: 400,
  fontVariantNumeric: 'tabular-nums',
  display: 'block',
  marginBlockStart: '0',
  marginBlockEnd: '0',
  marginInlineStart: '0',
  marginInlineEnd: '0',

  variants: {
    size: {
      tiny: {
        ...tinyStyles,
        '&::before': {
          ...tinyStyles['::before'],
        },
        '&::after': {
          ...tinyStyles['::after'],
        },
      },
      small: {
        ...smallStyles,
        '&::before': {
          ...smallStyles['::before'],
        },
        '&::after': {
          ...smallStyles['::after'],
        },
      },
      standard: {
        ...standardStyles,
        '&::before': {
          ...standardStyles['::before'],
        },
        '&::after': {
          ...standardStyles['::after'],
        },
      },
      large: {
        ...largeStyles,
        '&::before': {
          ...largeStyles['::before'],
        },
        '&::after': {
          ...largeStyles['::after'],
        },
      },
    },
  },
  compoundVariants: [],
  defaultVariants: {
    size: 'standard',
  },
});
