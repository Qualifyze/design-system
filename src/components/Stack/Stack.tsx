// *Heavily* inspired by (and copied from) Braid https://github.com/seek-oss/braid-design-system/blob/26ea250b1565a1d65099ff3298fd474dd6fd1a10/lib/components/Stack/Stack.tsx
import * as React from 'react';
import flattenChildren from 'react-keyed-flatten-children';
import { ReactNodeNoStrings } from '../private/ReactNodeNoStrings';
import { Box, BoxProps } from '../Box/Box';
import { useNegativeValue } from '../private/hooks/useNegativeValue';

const alignToDisplay = {
  left: 'block',
  center: 'flex',
  right: 'flex',
} as const;

export const validStackComponents = ['div', 'ol', 'ul'] as const;

// interface UseStackItemProps {
//   align: ResponsiveProp<Align>;
//   space: BoxProps['paddingTop'];
// }

export interface StackProps {
  as?: typeof validStackComponents[number];
  children: ReactNodeNoStrings;
  space: BoxProps['pt'];
  // align?: ResponsiveProp<Align>;
}

export const Stack = ({
  as = `div`,
  space = 0,
  children,
  align = `left`,
}: StackProps) => {
  const stackItems = flattenChildren(children);

  // when the stack should be a list we need to render `<li>`s
  const isList = as === 'ol' || as === 'ul';
  const stackItemComponent = isList ? 'li' : 'div';
  //  const stackItemProps = useStackItem({ space, align })

  return (
    <Box as={as} sx={{ mt: useNegativeValue(space) }}>
      {React.Children.map(stackItems, child => {
        return <Box as={stackItemComponent}>{child}</Box>;
      })}
    </Box>
  );
};
