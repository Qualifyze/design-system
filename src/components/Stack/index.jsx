/**
 * Heavily inspired by Braid https://github.com/seek-oss/braid-design-system/blob/26ea250b1565a1d65099ff3298fd474dd6fd1a10/lib/components/Stack/Stack.tsx
 */
import React, { Children } from 'react'
import PropTypes from 'prop-types'
import flattenChildren from 'react-keyed-flatten-children'

import { propType } from '../../util/style'
import Box from '../Box'

const useStackItem = ({ align }) => ({
  width: '100%',
  // If we're aligned left across all screen sizes,
  // there's actually no alignment work to do.
  ...(align === 'left'
    ? null
    : {
        display: 'flex',
        flexDirection: 'column',
        alignItems: align === 'center' ? 'center' : 'flex-end',
      }),
})

const Stack = ({ as, space, children, align }) => {
  const stackItems = flattenChildren(children)

  // when the stack should be a list we need to render `<li>`s
  const isList = as === 'ol' || as === 'ul'
  const stackItemComponent = isList ? 'li' : 'div'
  const stackItemProps = useStackItem({ align })

  return (
    <Box
      as={as}
      sx={{
        position: 'static',
        display: 'flex',
        flexDirection: 'column',
        gap: space,
      }}
    >
      {Children.map(stackItems, child => {
        return (
          <Box
            as={stackItemComponent}
            {...stackItemProps}
            sx={{ position: 'static' }}
          >
            {child}
          </Box>
        )
      })}
    </Box>
  )
}

Stack.defaultProps = {
  as: 'div',
  space: 0,
  align: 'left',
}

Stack.propTypes = {
  /** Sets the HTML tag for the component */
  as: PropTypes.oneOf(['div', 'ol', 'ul']),
  /** Child React components to render */
  children: PropTypes.node.isRequired,
  /** Spacing between the children */
  space: propType,
  /** How to align the text */
  align: PropTypes.oneOf(['left', 'center', 'right']),
}

export default Stack
