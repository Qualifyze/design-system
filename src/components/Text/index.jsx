/**
 * TODO
 *
 * @author Gregor Albrecht <hello@gregoralbrecht.com>
 */
import PropTypes from 'prop-types'
import shouldForwardProp from '@styled-system/should-forward-prop'

import { styled, sx, variant, space, propTypes } from '../../util/style'

const Text = styled('p', { shouldForwardProp })(
  props => ({
    '& b': {
      fontWeight: props.theme.fontWeights.semibold,
    },
  }),
  space,
  variant({
    prop: 'size',
    variants: {
      tiny: {
        fontSize: 0,
      },
      small: {
        fontSize: 1,
      },
      standard: {
        fontSize: 2,
      },
      large: {
        fontSize: 3,
      },
    },
  }),
  variant({
    prop: 'weight',
    variants: {
      regular: {
        fontWeight: 200,
      },
      medium: {
        fontWeight: 400,
      },
      strong: {
        fontWeight: 500,
      },
    },
  }),
  variant({
    prop: 'tone',
    variants: {
      neutral: {
        color: 'grey.800',
      },
      lightneutral: {
        color: 'hsla(0, 0%, 100%, 1)',
      },
      positive: {
        color: 'green.700',
      },
      critical: {
        color: 'red.700',
      },
      secondary: {
        color: 'grey.800',
      },
      lightsecondary: {
        color: 'hsla(0, 0%, 100%, 0.6)',
      },
      info: {
        color: 'primary.600',
      },
    },
  }),
  variant({
    prop: 'align',
    variants: {
      left: {
        textAlign: 'left',
      },
      center: {
        textAlign: 'center',
      },
      right: {
        textAlign: 'right',
      },
    },
  }),
  sx
)

Text.displayName = 'Text'

Text.defaultProps = {
  as: 'p',
  size: 'standard',
  align: 'left',
  tone: 'neutral',
}

Text.propTypes = {
  /** Sets the HTML tag for the component */
  as: PropTypes.string,
  /** The font size */
  size: PropTypes.oneOf(['tiny', 'small', 'standard', 'large']),
  /** How to align the text */
  align: PropTypes.oneOf(['left', 'center', 'right']),
  /** Tone, i.e., color to use */
  tone: PropTypes.oneOf([
    'neutral',
    'lightneutral',
    'positive',
    'critical',
    'secondary',
    'lightsecondary',
    'info',
  ]),
  ...propTypes.space,
}

export default Text
