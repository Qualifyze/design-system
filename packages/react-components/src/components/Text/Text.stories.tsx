import React from 'react';
import {
  DownloadIcon,
  GearIcon,
  CheckIcon,
  Cross2Icon,
} from '@radix-ui/react-icons';

import { Flex } from '../Flex/Flex';

import { Text } from './Text';

export default {
  title: 'Components/Text',
  component: Text,
};

export const basic = () => (
  <>
    <Text size="tiny">Hola mundo</Text>
    <Text size="tiny">Hola mundo</Text>
    <Text size="tiny">Hola mundo</Text>
    <Text size="small">Hola mundo</Text>
    <Text size="small">Hola mundo</Text>
    <Text size="small">Hola mundo</Text>
    <Text size="standard">Hola mundo</Text>
    <Text size="standard">Hola mundo</Text>
    <Text size="standard">Hola mundo</Text>
    <Text size="large">Hola mundo</Text>
    <Text size="large">Hola mundo</Text>
    <Text size="large">Hola mundo</Text>
  </>
);

export const paragraphs = () => (
  <>
    <Text>
      Pivot termsheet twitter. Leverage deployment holy grail advisor traction
      monetization seed money business model canvas validation ownership pitch
      infographic. Conversion direct mailing social proof release validation
      disruptive termsheet alpha leverage long tail hackathon iteration user
      experience. Hypotheses advisor paradigm shift hackathon interaction design
      ecosystem sales research & development business plan iteration.
    </Text>
    <Text>
      Equity startup alpha hypotheses accelerator iPad. Gamification holy grail
      MVP business model canvas funding pitch responsive web design mass market
      paradigm shift virality. Termsheet customer user experience return on
      investment metrics interaction design validation hackathon social media
      direct mailing client funding value proposition business plan. Disruptive
      market low hanging fruit prototype paradigm shift.
    </Text>
    <Text>
      Branding focus buzz niche market handshake android virality user
      experience. MVP crowdfunding incubator first mover advantage strategy
      disruptive hypotheses founders long tail focus venture.
    </Text>
  </>
);

export const withIcon = () => (
  <Flex css={{ fd: 'column', gap: '$32' }}>
    <Text size="tiny">
      <DownloadIcon />
      Hola mundo
    </Text>
    <Text size="small">
      <GearIcon />
      Hola mundo
    </Text>
    <Text size="standard">
      <CheckIcon />
      Hola mundo
    </Text>
    <Text size="large">
      <Cross2Icon />
      Hola mundo
    </Text>
  </Flex>
);
