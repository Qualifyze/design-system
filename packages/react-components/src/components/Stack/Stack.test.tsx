import * as React from 'react';
import { render, screen } from '@testing-library/react';

import { Text } from '../Text/Text';

import { Stack } from './Stack';

describe('Stack', () => {
  test('renders', () => {
    render(
      <Stack>
        <Text>hola</Text>
        <Text>hello</Text>
      </Stack>
    );
    expect(screen.getByText(/hola/)).toBeInTheDocument();
  });

  test('renders with theme-aware gap prop', () => {
    render(
      <Stack gap="32">
        <Text>hola</Text>
        <Text>hello</Text>
      </Stack>
    );
    const B = screen.getByText(/hola/);
    // The `toHaveStyle` matcher relies on a serializer of the created CSS,
    // which we don't have for Stitches yet
    // expect(B).toHaveStyle(`padding-top`, `1px`);
    expect(B).toMatchSnapshot();
  });

  test('passes `data-` attributes through to the DOM', () => {
    render(<Stack data-testid="some-random-id">hola</Stack>);
    const B = screen.getByText(/hola/);
    expect(B).toHaveAttribute(`data-testid`, `some-random-id`);
    expect(B).toMatchSnapshot();
  });

  test('passes `aria-` attributes through to the DOM', () => {
    render(<Stack aria-hidden="true">hola</Stack>);
    const B = screen.getByText(/hola/);
    expect(B).toHaveAttribute(`aria-hidden`, `true`);
    expect(B).toMatchSnapshot();
  });

  test('renders as passed HTML element', () => {
    render(
      <Stack as="ul">
        <Text>hola</Text>
        <Text>hello</Text>
      </Stack>
    );
    const B = screen.getByRole(`list`);
    expect(B).toMatchSnapshot();
  });
});
