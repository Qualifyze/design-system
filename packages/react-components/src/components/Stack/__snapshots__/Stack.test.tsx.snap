// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack passes \`aria-\` attributes through to the DOM 1`] = `
<div
  aria-hidden="true"
  class="c-gLVhRE c-gLVhRE-gjemnU-align-start c-gLVhRE-etXRKT-gap-24"
>
  hola
</div>
`;

exports[`Stack passes \`data-\` attributes through to the DOM 1`] = `
<div
  class="c-gLVhRE c-gLVhRE-gjemnU-align-start c-gLVhRE-etXRKT-gap-24"
  data-testid="some-random-id"
>
  hola
</div>
`;

exports[`Stack renders as passed HTML element 1`] = `
<ul
  class="c-gLVhRE c-gLVhRE-gjemnU-align-start c-gLVhRE-etXRKT-gap-24"
>
  <p
    class="c-kVPbLv c-kVPbLv-bpkWGi-size-standard"
  >
    hola
  </p>
  <p
    class="c-kVPbLv c-kVPbLv-bpkWGi-size-standard"
  >
    hello
  </p>
</ul>
`;

exports[`Stack renders with theme-aware gap prop 1`] = `
<p
  class="c-kVPbLv c-kVPbLv-bpkWGi-size-standard"
>
  hola
</p>
`;
